pkg/kn/commands/root.go

```diff
@@ -16,19 +16,44 @@ package commands
 
 import (
 	"fmt"
+	"io"
 	"os"
 	"path/filepath"
 
+	serving "github.com/knative/serving/pkg/client/clientset/versioned/typed/serving/v1alpha1"
 	homedir "github.com/mitchellh/go-homedir"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 var cfgFile string
 var kubeCfgFile string
 
+// Parameters for creating commands. Useful for inserting mocks for testing.
+type KnParams struct {
+	Output         io.Writer
+	ServingFactory func() (serving.ServingV1alpha1Interface, error)
+}
+
+func (c *KnParams) Initialize() {
+	if c.ServingFactory == nil {
+		c.ServingFactory = GetConfig
+	}
+}
+
 // rootCmd represents the base command when called without any subcommands
-func NewKnCommand() *cobra.Command {
+func NewKnCommand(params ...KnParams) *cobra.Command {
```

Comment:
So that NewKnCommand can have no parameters. Following the practice detailed [here](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis), which I liked for the entry we expect people to use to a lib.

Reply:
NB: Open to going to the full "functional options" route too, but this seemed a reasonable compromise between easy default API and simple configuration API.

Reply:
k, lets keep it this way for now.